Add protocol option to all queries

fsuste opened the pull request:
## Pull Request Checklist

**Is this in reference to an existing issue?**

#### General

- [x] Update Changelog following the conventions laid out on [Keep A Changelog](http://keepachangelog.com/)

- [x] Update README with any necessary configuration snippets

- [x] Binstubs are created if needed

- [x] RuboCop passes

- [x] Existing tests pass

#### New Plugins

- [x] Tests

- [x] Add the plugin to the README

- [x] Does it have a complete header as outlined [here](http://sensu-plugins.io/docs/developer_guidelines.html#coding-style)

#### Purpose

#### Known Compatibility Issues

fsuste committed:
Add protocol option to all queries

fsuste committed:
Update changelog

fsuste committed:
Fix rubocop errors

fsuste commented:
This is a pull request resolving issue #8

majormoses commented:
Thanks for your contribution to Sensu plugins! Without people like you submitting PRs we couldn't run the project. I will review it shortly.

majormoses commented on CHANGELOG.md:
```diff
@@ -4,6 +4,8 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 This CHANGELOG follows the format listed at [Keep A Changelog](http://keepachangelog.com/)
 
 ## [Unreleased]
+### Changed
```
This should be `### Added` as that is a new feature the ability to change the protocol.

majormoses commented on CHANGELOG.md:
```diff
@@ -4,6 +4,8 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 
 ## [Unreleased]
+### Changed
+- Add protocol option which can be passed asn an argument to specify which protocol to use
```
> asn an argument

looks like a typo

majormoses commented on bin/check-graylog-buffers.rb:
```diff
@@ -37,6 +37,12 @@
 class CheckGraylogBuffers < Sensu::Plugin::Check::CLI
+  option :protocol,
+         description: 'Protocol for connecting to Graylog',
+         short: '-proto',
```
I am actually not sure if short is ok with multi char short args. Have you tested this? Usually short options single chars that can be chained together like `tar -xvf` that is not a single arg but actually 3 of them.

majormoses commented on bin/check-graylog-buffers.rb:
```diff
@@ -37,6 +37,12 @@
+         description: 'Protocol for connecting to Graylog',
+         short: '-proto',
+         long: '--protocol PROTOCOL',
+         default: 'http'
```
As much as I feel `https` should be the default I agree that the breaking change would be unwarranted right now.

majormoses reviewed (changes requested):
Overall this looks great one minor changelog tweak and a [testing artifact](https://github.com/sensu-plugins/community/blob/master/PULL_REQUEST_PROCESS.md#7-testing-artifacts) is all I need and to get this merged and released.

fsuste committed:
Fix requested changes

fsuste commented:
I would also like to see `https` as a default, but that is an expensive change version wise... I am now working on a testing artifact.

fsuste commented:
I don't actually see how can I test this except for unit testing, but that would be testing the mixlib-cli gem... If you have a better idea for this, I would love to hear it :)

majormoses commented:
Well there are a couple options:
- unit testing
- integration testing
- manual testing

I would suggest you read the following articles where I outline the types of testing and go in depth with integration testing:
- https://blog.sensuapp.org/our-approach-to-testing-sensu-plugins-869260eba94c
- https://blog.sensuapp.org/writing-sensu-plugin-tests-with-test-kitchen-and-serverspec-b646d2eeee51

The easiest is a manual test, just run the check you would have run via sensu in cli and capture the input and output in a gist.

nathanhruby commented:
The tests in check-graylog-streams_spec.rb could be used as an example for checking the args and stubbing out the actual call to make sure it is used as intended.

majormoses commented on bin/check-graylog2-alive.rb:
```diff
@@ -37,6 +37,12 @@
 class CheckGraylog2Alive < Sensu::Plugin::Check::CLI
+  option :protocol,
+         description: 'Protocol for connecting to Graylog',
+         short: '-s',
```
let's drop the `-s` as it no longer makes sense and `-p` is usually `port`

majormoses commented on bin/check-graylog-streams.rb:
```diff
@@ -26,6 +26,12 @@
 class CheckGraylogStreams < Sensu::Plugin::Check::CLI
+  option :protocol,
+         description: 'Protocol for connecting to Graylog',
+         short: '-s',
```
let's drop the `-s` as it no longer makes sense and `-p` is usually `port`

majormoses commented on bin/check-graylog-buffers.rb:
```diff
@@ -37,6 +37,12 @@
 class CheckGraylogBuffers < Sensu::Plugin::Check::CLI
+  option :protocol,
+         description: 'Protocol for connecting to Graylog',
+         short: '-s',
```
let's drop the `-s` as it no longer makes sense and `-p` is usually `port`

majormoses commented on bin/metrics-graylog.rb:
```diff
@@ -38,6 +38,12 @@
 class MetricsGraylog < Sensu::Plugin::Metric::CLI::Graphite
+  option :protocol,
+         description: 'Protocol for connecting to Graylog',
+         short: '-s',
```
let's drop the `-s` as it no longer makes sense and `-p` is usually `port`

majormoses reviewed (changes requested):
A few more things I noticed that could be improved:
- drop the `-s` option as the name does not make sense and `-p` is used for `port`
- specify the available options for `protocol` like [this](https://github.com/sensu-plugins/sensu-plugins-redis/blob/2.3.2/lib/redis_client_options.rb#L75)

majormoses committed:
remove the `-s` short option as it does not make sense for protocol.

restricted the input of protocol to be `http` or `https`.

Signed-off-by: Ben Abrams <me@benabrams.it>

majormoses commented:
I have made the self requested modifications if someone can provide either a manual or automated testing artifact I am :+1: to merge this.

majormoses committed:
adding support for making `insecure` requests.

While a much better answer is to install your ca-certs to ensure this is not needed this enables easier local development. I would not reccomend using this in production and suggest using something like let's encrypt, a well known CA, or a service like aws certificate manager.

Signed-off-by: Ben Abrams <me@benabrams.it>

majormoses commented:
@nathanhruby @Narayanan170  any chance you have a graylog that we can test against? I don't see how this PR is the culprit of https://github.com/sensu-plugins/sensu-plugins-graylog/issues/8#issuecomment-365294851 but it does mean we need to ensure that we have a test before I am OK with merging this.

nathanhruby commented:
I don't work at a place with graylog anymore, but I might have some time later this week to spin up a test cluster in a docker as test of something else I am doing.

majormoses commented:
I can confirm this works:
```
$ bundle exec ./bin/check-graylog2-alive.rb --protocol https --host graylog.$DOMAIN.$TLD --username babrams --password $(cat creds) --port 443 --apipath /api
CheckGraylog2Alive OK: Graylog2 server is: running/true/alive
```

majormoses commented:
released: https://rubygems.org/gems/sensu-plugins-graylog/versions/1.3.0
